### PR TITLE
168524581: Authorization ADR

### PR DESCRIPTION
**Pull request by jim**
## Description

This PR includes the role-based auth ADR that has been simmering for a while now.

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/169736889) for this change

### CODE REVIEW

File: .spelling [hunk 1]
```diff
@@ -520,7 +520,28 @@ Fargate
...
+ - docs/frontend.md
+ - ./README.md
+ - ./docs/data/tspp-data-creation.md
+ - ./docs/data/tariff400ng-yearly-import.md
+ - ./scripts/README.md
+ - ./docs/how-to/run-storybook.md
+ - ./docs/database.md
+ - ./docs/._database.md
+ - ./docs/how-to/test-virus-scanning.md
+ - ./docs/adr/0039-role-base-authorization.md
```
Comment (chrisgilmerproj) [inline .spelling]: Does this just ignore all the files here?

File: docs/adr/0040-role-base-authorization.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
+# Add Role-Based Authorization
+
+**NOTE:** This ADR updates and supersedes [ADR0024 Model Authorization and Handler Design](./0024-model-authorization-and-handler-design.md).
```
Comment (chrisgilmerproj) [inline docs/adr/0040-role-base-authorization.md]: Can we add a note to that ADR as well?

File: .spelling (hunk 1 above)
Comment (jim) [reply .spelling]: Yeah, this was a bad merge. I'm not actually sure what happened here. Will fix.

Comment (jim) [reply .spelling]: Like, how did `/usr/local` get in there?

File: docs/adr/0040-role-base-authorization.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+## Considered Alternatives
+
+* **Implement Role-Based Access Control in Middleware**
+
+  Swagger yaml files include a new ‘roles’ tag on each endpoint,
+indicating the list of roles that can access the endpoint.
+The `UserAuthMiddleware` is edited to compare this list with the roles the user has
+(by checking the session).
+
+  We expect that there will be database changes made to allow for there to be a many-to-many relationship between users and roles.
```
Comment (chrisgilmerproj) [inline docs/adr/0040-role-base-authorization.md]: Is this the `users` table or each of the individual `*_users` tables?

File: docs/adr/0040-role-base-authorization.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+with that PPM is indeed the one making the request. An error is returned if not.
+Note that this process takes only one database call.
+
+  We can continue using this pattern as we expand the MilMove system.
+
+* **Create an Enforcer Object**
+
+  This approach involves writing code to encapsulate the authorization logic and explicitly invoking it within handlers or service objects as needed. It could be seen as taking our existing system and just extracting the authorization checks into functions outside the models.
+
+  The name for this object comes from what similar objects are called within casbin.
```
Comment (chrisgilmerproj) [inline docs/adr/0040-role-base-authorization.md]: Is the downside here that if we fail to add an enforcer to a handler then its exposed to use with no checks?

Comment (chrisgilmerproj) [reply docs/adr/0040-role-base-authorization.md]: nm, I see it below.

File: docs/adr/0040-role-base-authorization.md (hunk 2 above)
Comment (tinyels) [reply docs/adr/0040-role-base-authorization.md]: That's a good idea @chrisgilmerproj !

File: docs/adr/0040-role-base-authorization.md (hunk 3 above)
Comment (tinyels) [reply docs/adr/0040-role-base-authorization.md]: I think it is fine to defer details of the database changes for later work. I would like to make sure that all of the teams (and I)  get a chance to review the PRs for that work though.

File: docs/adr/0040-role-base-authorization.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  * It allows us to easily audit which endpoints are accessible by which roles by examining the API descriptions in the swaggerfile.
+  * Having the role check in a middleware makes it less likely that an authorization check is missed.
+  * We have completed spikes that show that this approach is doable with a minimal amount of additional work.
+
+## Pros and Cons of the Alternatives
+
+### *Implement Role-Based Access Control in Middleware*
+
+* `+` Role/Endpoint associations live in Swagger yaml file, which is easy to read/modify
+* `+` Flexible to addition of new role types
```
Comment (tinyels) [inline docs/adr/0040-role-base-authorization.md]: Another pro is that having the roles in the documentation will help communicate that.

**Pull request merged**